### PR TITLE
Fix/issue 2128

### LINKED ISSUE #2128: Unit requires different MPI flavor from configured
Task description :
```
t.name = 'Preprocessing'       
t.executable = 'python'         
t.arguments = ['~/iwp_preprocessing.py'] 
t.cpu_reqs = {'processes': 28,
              'thread_type': 'OpenMP', 
              'threads_per_process': 1,
              'process_type': 'MPI'}
```

RADICAL-Stack
```
  python               : 3.7.0
  pythonpath           :
  virtualenv           : /home/aymen/iwp

  radical.entk         : 1.0.2
  radical.pilot        : 1.3.0
  radical.saga         : 1.3.0
  radical.utils        : 1.3.1
```

Pre_bootstrap loaded modules:
```
"module purge",
"module load gcc",
"module load mpi/gcc_openmpi",
"module load slurm",
"module load anaconda3
```
# Pre-exec commands and modules to load:
"module unload intel/19.5"
"module load intel/18.4"
"
In `unit000000.sh`  under `# The command to run`  my command look like the following:

```
mpirun  -np 28  -host r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351,r351 -x "LD_LIBRARY_PATH" -x "PATH" -x "PYTHONPATH" -x "OMP_NUM_THREADS" -x "CUDA_VISIBLE_DEVICES" -x "RP_AGENT_ID" -x "RP_GTOD" -x "RP_PILOT_ID" -x "RP_PILOT_STAGING" -x "RP_PROF" -x "RP_SESSION_ID" -x "RP_SPAWNER_ID" -x "RP_TMP" -x "RP_UNIT_ID" -x "RP_UNIT_NAME" -x "CUDA_VISIBLE_DEVICES"  python "~/iwp_preprocessing.py" 
```

The `STDERR` : 

```
[mpiexec@r351.pvt.bridges.psc.edu] match_arg (../../utils/args/args.c:254): unrecognized argument x
[mpiexec@r351.pvt.bridges.psc.edu] HYDU_parse_array (../../utils/args/args.c:269): argument matching returned error
[mpiexec@r351.pvt.bridges.psc.edu] parse_args (../../ui/mpich/utils.c:4773): error parsing input array
[mpiexec@r351.pvt.bridges.psc.edu] HYD_uii_mpx_get_parameters (../../ui/mpich/utils.c:5109): unable to parse user arguments
```

Based on Andre's preliminary analysis this is might happen because of :
```The switch between Hydra and Orte based MPIs is not working```

Next steps: 

- [x] Find out how `environment variables` are set in that version of `mpirun`
- [x] Check (https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/agent/launch_method/mpirun.py#L101)
- [x] Check ( https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/agent/launch_method/base.py#L369)

### PR DESCRIPTION
**Pull request by andre-merzky**
This fixes #2128 by checking for intel mpi.

### CODE REVIEW

**Comment by AymenFJA** [comment]
Testing now! with the same configurations based on #2128.

**Comment by AymenFJA** [comment]
This issue, still exists. The `construct_command` function is failing to build the command based on the received `MPI_FLAVOR`,  I am investigating more.

**Comment by andre-merzky** [comment]
@AymenFJA : Hey - can you pick this up again, please, and let us know if (a) this is still an issue, and (b) if you need help on the fix? Thanks!

**Comment by andre-merzky** [comment]
ping @AymenFJA

**Comment by AymenFJA** [comment]
> ping @AymenFJA

Hey @andre-merzky . I apologize for the late response I will pick this up!

**Comment by andre-merzky** [comment]
Great, thank you!

**Comment by iparask** [comment]
@andre-merzky, can you also update the tests here? There is a base class test for the launch methods. Thank you!

**Comment by andre-merzky** [comment]
> @andre-merzky, can you also update the tests here? There is a base class test for the launch methods. Thank you!

Let me delegate the test update to @AymenFJA , please - this PR is owned by him now...

**Comment by AymenFJA** [comment]
> > @andre-merzky, can you also update the tests here? There is a base class test for the launch methods. Thank you!
> 
> Let me delegate the test update to @AymenFJA , please - this PR is owned by him now...

I will do that, thanks Andre.

**Comment by AymenFJA** [comment]
@andre-merzky So far, I did not face issue #2128 while testing the PR.  
Also, I have extended the unit test for this issue. Also, while I was trying to test different `MPI` implementations `RP` failed with `MVAPICH `. I do understand that it is not supported yet in `RP` so I have created the test for `MVAPICH` and created a warning for future reference in case we want to support that flavor. Please let me know if there is anything else I should do.

File: tests/test_lm/test_unit/test_base.py [hunk 1]
```diff
@@ -58,3 +59,23 @@ def test_get_mpi_info(mocked_init):
...
+    ru.sh_callout.side_effect = [['test',1,1],['Intel(R) MPI Library for Linux* OS,\n\n\
+                                               Version 2019 Update 5 Build 20190806\n\n\
+                                               Copyright 2003-2019, Intel Corporation.',3,0]]
+    version, flavor = lm._get_mpi_info('mpirun')
+    assert version == ''
+    assert flavor == 'HYDRA'
+
+    ru.sh_callout.side_effect = [['test',1,1],['MVAPICH2 2.3b',3,0]]
+    version, flavor = lm._get_mpi_info('mpirun')
+    try:
```
Comment (iparask) [inline tests/test_lm/test_unit/test_base.py:77]: @AymenFJA which resource has this MPI version?

Comment (AymenFJA) [reply tests/test_lm/test_unit/test_base.py:77]: XSEDE Bridges.

**Comment by andre-merzky** [comment]
> while I was trying to test different `MPI` implementations `RP` failed with `MVAPICH`

How did it fail for MVAPICH?

**Comment by AymenFJA** [comment]
> > while I was trying to test different `MPI` implementations `RP` failed with `MVAPICH`
> 
> How did it fail for MVAPICH?

Ooops, actually it is not. Giannis and I tried it and it just now and it is working just as expected. But, yesterday I had a bunch of errors from running `mpirun --version` using `mpi/gcc_mvapich`. I will fix the test to match the case.

**Comment by AymenFJA** [comment]
@iparask, ready to move on with this PR.

File: tests/test_lm/test_unit/test_base.py [hunk 2]
```diff
@@ -13,48 +14,55 @@
...
+
+        lm = LaunchMethod(name=None, cfg={}, session=None)
+        lm._log = mock.Mock()
+        ru.sh_callout = mock.Mock()
+        ru.sh_callout.side_effect = [['test',1,0]]
+        version, flavor = lm._get_mpi_info('mpirun')
+        if version is None:
+            assert True
+        else:
+            assert False
```
Comment (iparask) [inline tests/test_lm/test_unit/test_base.py:40]:
```suggestion
self.assertIsNone(version)
```

I think this should do the trick.

```

**Review by iparask — CHANGES_REQUESTED** [review]
If the one comment I have is resolved, I am happy with the PR.

As soon as the comment is resolved, I think it can be merged.

File: tests/test_lm/test_unit/test_base.py (hunk 2 above)
Comment (andre-merzky) [reply tests/test_lm/test_unit/test_base.py:40]: or
```py
assert(version is None)
```

Comment (AymenFJA) [reply tests/test_lm/test_unit/test_base.py]: Done. The requested change was committed.

**Review by iparask — APPROVED** [review]
Looks good to me

**Pull request merged**